File: 132__Palindrome_Partitioning_II.py
class Solution(object):
    def minCut(self, s):
        """
        :type s: str
        :rtype: int
        """
        n = len(s)
        dp = [0] * n
        pal = [[False] * n for row in range(n)]

        for j in range(n):
            min_cut = n
            for i in range(j + 1):
                if s[i] == s[j] and (j - i <= 2 or pal[i + 1][j - 1]):
                    pal[i][j] = True
                    if i == 0:
                        min_cut = 0
                    else:
                        min_cut = min(min_cut, dp[i - 1] + 1)
            dp[j] = min_cut

        return dp[-1]

File: Java/132__Palindrome_Partitioning_II.java
class Solution {
    public int minCut(String s) {
        char[] s_arr = s.toCharArray();
        int len = s.length();
        boolean isPalindrome[][] = new boolean[len][len];
        int dp[] = new int[len];

        for (int end = 0; end < len; end++) {
            int min = Integer.MAX_VALUE;
            for (int start = 0; start <= end; start++) {
                if (s_arr[end] == s_arr[start] && (start + 1 >= end || isPalindrome[start + 1][end - 1]) ){
                    isPalindrome[start][end] = true;
                    min = start == 0 ? 0 : Math.min(min, dp[start - 1] + 1);
                }
            }
            dp[end] = min;
        }
        return dp[len - 1];
    }
}
